### PR TITLE
test: Add cache correctness test

### PR DESCRIPTION
**Pull request by salvacorts**
Follwup to https://github.com/grafana/loki/pull/20986 adding the correctness cache we used to triage the implementation.

### CODE REVIEW

File: pkg/logql/bench/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+		require.Equal(t, rt1, rt2, "resultType differs for query %q", r.tc.Query)
+		require.JSONEq(t, result1, result2, "cached result differs for query %q", r.tc.Query)
+	}
+
+	// --- Log cache: only caches empty results ---
+	// Identify queries that are guaranteed to return empty results.
+	const impossibleFilter = `|= "this will not hit any line"`
+	var logEmptyCases []TestCase
+	for _, tc := range cases {
+		if tc.Kind() == "log" && strings.Contains(tc.Query, impossibleFilter) {
```
Comment (sophiewaldman) [inline pkg/logql/bench/cache_test.go]: Can we maybe use the tags here instead of hard-coding a match against the query?

Comment (salvacorts) [reply pkg/logql/bench/cache_test.go]: good call. I exposed the tags in the TestCase and used it. I also fixed a test that was broken due to the empty response.

File: pkg/logql/bench/bench_test.go [hunk 2]
```diff
@@ -541,6 +546,29 @@ func assertResultNotEmpty(t *testing.T, data parser.Value, message string) {
...
+		isEmpty := true
+		for _, series := range v {
+			if len(series.Floats) > 0 || len(series.Histograms) > 0 {
+				isEmpty = false
+				break
+			}
+		}
+		require.True(t, isEmpty, message+" - matrix has series with data points")
+	case promql.Scalar:
+		// Scalars always have a value, so no need to check
```
Comment (sophiewaldman) [inline pkg/logql/bench/bench_test.go]: Should we return false in the case where we're returning a scalar?

File: pkg/logql/bench/bench_test.go [hunk 3]
```diff
@@ -541,6 +546,29 @@ func assertResultNotEmpty(t *testing.T, data parser.Value, message string) {
 	}
 }
 
+func assertResultEmpty(t *testing.T, data parser.Value, message string) {
+	t.Helper()
+	switch v := data.(type) {
+	case promql.Vector:
+		require.Empty(t, v, message)
+	case promql.Matrix:
+		isEmpty := true
```
Comment (sophiewaldman) [inline pkg/logql/bench/bench_test.go:556]: I see in `assertResultNotEmpty` we still call `require.notEmpty(t, v, message)` for `promql.Matrix` - do we want to call `require.Empty` here?

Comment (salvacorts) [reply pkg/logql/bench/bench_test.go:556]: Done
